src/System.Private.CoreLib/shared/System/Collections/Concurrent/ConcurrentQueueSegment.cs

```diff
@@ -112,15 +112,13 @@ internal static int RoundUpToPowerOf2(int i)
                 _frozenForEnqueues = true;
 
                 // Increase the tail by FreezeOffset, spinning until we're successful in doing so.
-                var spinner = new SpinWait();
                 while (true)
                 {
                     int tail = Volatile.Read(ref _headAndTail.Tail);
                     if (Interlocked.CompareExchange(ref _headAndTail.Tail, tail + FreezeOffset, tail) == tail)
```

Comment:
Should this loop also use the result of the `Interlocked.CompareExchange` rather than doing another `Volatile.Read`?
e.g.
```csharp
int tail = Volatile.Read(ref _headAndTail.Tail);
while (true)
{
    int prev = Interlocked.CompareExchange(ref _headAndTail.Tail, tail + FreezeOffset, tail);
    if (prev == tail)
    {
        break;
    }
    tail = prev;
}
```

Reply:
Since there is no pause any longer so the return should be fresh

Reply:
Looks to be same with other 3 loops in this PR; a reread is needed with the spin (as the return may not be up-to date after the pause). 

Without it the CompareExchange return value can be used as the reread; then the loop stays in registers rather than accessing memory again.